VxWorks/Os/Mutex.cpp

```diff
@@ -5,28 +5,37 @@
 #include "VxWorks/Os/Mutex.hpp"
 #include <Fw/Types/Assert.hpp>
 #include "VxWorks/Os/error.hpp"
-
 namespace Os {
 namespace VxWorks {
 namespace Mutex {
 
 VxWorksMutex::VxWorksMutex() : Os::MutexInterface(), m_handle() {
     this->m_handle.m_mutex_descriptor = semMCreate(SEM_Q_PRIORITY | SEM_INVERSION_SAFE | SEM_DELETE_SAFE);
-    FW_ASSERT(this->m_handle.m_mutex_descriptor != SEM_ID_NULL);
+    FW_ASSERT(this->m_handle.m_mutex_descriptor != SEM_ID_NULL, static_cast<FwAssertArgType>(errno));
 }
 
 VxWorksMutex::~VxWorksMutex() {
-    (void)semDelete(this->m_handle.m_mutex_descriptor);
+    STATUS status = semDelete(this->m_handle.m_mutex_descriptor);
+    FW_ASSERT(status == VXWORKS_OK, static_cast<FwAssertArgType>(errno));
 }
 
 VxWorksMutex::Status VxWorksMutex::take() {
     STATUS status = semTake(this->m_handle.m_mutex_descriptor, WAIT_FOREVER);
-    return Os::VxWorks::vxworks_status_to_mutex_status(status);
+    PlatformIntType statusType = (status == VXWORKS_OK) ? VXWORKS_OK : errno;
+
+    // Keep track of semTakes and assert we are not re-entering.
+    this->m_handle.m_sem_take_counter++;
+    FW_ASSERT(this->m_handle.m_sem_take_counter == 1, static_cast<FwAssertArgType>(this->m_handle.m_sem_take_counter));
+
+    return Os::VxWorks::vxworks_status_to_mutex_status(statusType);
 }
 
 VxWorksMutex::Status VxWorksMutex::release() {
+    // Keep track of semTakes
+    this->m_handle.m_sem_take_counter--;
```

Comment:
Probably should document this rentrant patch stuff. Alternatively, we might choose to switch to the binary sem which I believe is not inherently re-entrant.

Reply:
Good point. We'll add an SDD and capture this in that SDD.